opentelemetry-appender-tracing/src/layer.rs

```diff
@@ -1,61 +1,130 @@
-use opentelemetry::logs::{LogRecord, Logger, LoggerProvider, Severity};
+use opentelemetry::{
+    logs::{AnyValue, LogRecord, Logger, LoggerProvider, Severity},
+    Key,
+};
 use std::borrow::Cow;
-use tracing_core::Level;
+use tracing_core::{Level, Metadata};
+#[cfg(feature = "experimental_metadata_attributes")]
+use tracing_log::NormalizeEvent;
 use tracing_subscriber::Layer;
 
 const INSTRUMENTATION_LIBRARY_NAME: &str = "opentelemetry-appender-tracing";
 
 /// Visitor to record the fields from the event record.
-struct EventVisitor<'a> {
-    log_record: &'a mut LogRecord,
+#[derive(Default)]
+struct EventVisitor {
+    log_record_attributes: Vec<(Key, AnyValue)>,
+    log_record_body: Option<AnyValue>,
 }
 
-impl<'a> tracing::field::Visit for EventVisitor<'a> {
+#[cfg(feature = "experimental_metadata_attributes")]
+fn is_metadata(field: &'static str) -> bool {
+    field
+        .strip_prefix("log.")
+        .map(|remainder| {
+            matches!(
+                remainder,
+                "file"
+                    | "line"
+                    | "module_path"
+                    | "module.path"
+                    | "name"
+                    | "source.file.line"
+                    | "source.file.path"
+                    | "source.file.name"
+                    | "target"
+            )
+        })
+        .unwrap_or(false)
+}
+
+#[cfg(feature = "experimental_metadata_attributes")]
+fn get_filename(filepath: &str) -> &str {
+    if let Some((_, filename)) = filepath.rsplit_once('/') {
+        return filename;
+    }
+    if let Some((_, filename)) = filepath.rsplit_once('\\') {
+        return filename;
+    }
+    filepath
+}
+
+impl EventVisitor {
+    fn visit_metadata(&mut self, meta: &Metadata) {
+        self.log_record_attributes
+            .push(("name".into(), meta.name().into()));
+
+        #[cfg(feature = "experimental_metadata_attributes")]
+        self.log_record_attributes
+            .push(("log.target".into(), meta.target().to_owned().into()));
+
+        #[cfg(feature = "experimental_metadata_attributes")]
+        if let Some(module_path) = meta.module_path() {
+            self.log_record_attributes
+                .push(("log.module.path".into(), module_path.to_owned().into()));
+        }
+        #[cfg(feature = "experimental_metadata_attributes")]
+        if let Some(filepath) = meta.file() {
+            self.log_record_attributes
+                .push(("log.source.file.path".into(), filepath.to_owned().into()));
+            self.log_record_attributes.push((
+                "log.source.file.name".into(),
+                get_filename(filepath).to_owned().into(),
+            ));
+        }
+        #[cfg(feature = "experimental_metadata_attributes")]
+        if let Some(line) = meta.line() {
+            self.log_record_attributes
+                .push(("log.source.file.line".into(), line.into()));
+        }
```

Comment:
Can we wrap this into a function and annotate it with `#[cfg(feature = "experimental_metadata_attributes")]` once?

Reply:
Done